### ISSUE TITLE
Cosign with AWS KMS and AWS ECR doesn't work as expected for specific type of keys

### ISSUE THREAD

**Issue by beingamarnath** [sigstore/cosign#946]
### Description

Cosign has the support for AWS KMS as a means to provide keys for Signing and also AWS ECR is as well supported as a Container registry to have the signing information. 

So I tried to take the duo for testing and below are my findings .

P.S: I tried all the below setup also with Jfrog Artifactory as Container Registry and the results are the same.

AWS KMS 1 (alias/cosign-ec512):
-------------------------------
Key Type - Asymmetric
Origin: AWS_KMS
Key Spec: ECC_NIST_P521
Key Usage: Sign and Verify
Signing algorithms: ECDSA_SHA_512

Sign:
```
cosign sign --key awskms:///alias/cosign-ec512 <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:ECDSA_SHA_512
```
Sign Output:
 ```
 Pushing signature to: <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:sha256-6b444afd4dbdb32db7f5e31c7b70788680ff7e0876cf60e8acfbaa4b60740d15.sig
 ```
 Get Public Key:
 ```
 cosign public-key --key awskms:///alias/cosign-ec512 > kms-ec512.pub
 ```
 Verify:
 ```
cosign verify --key ./kms-ec512.pub <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:ECDSA_SHA_512
 ```
 Verify Output:
 ```
error: no matching signatures:
failed to verify signature
 ```

AWS KMS 2 (alias/cosign-ec384):
-------------------------------
Key Type - Asymmetric
Origin: AWS_KMS
Key Spec: ECC_NIST_P384
Key Usage: Sign and Verify
Signing algorithms: ECDSA_SHA_384

Sign:
```
cosign sign --key awskms:///alias/cosign-ec384 <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:ECDSA_SHA_384
```
Sign Output:
 ```
 Pushing signature to: <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>sha256-42823a9e45dccc30eed1c819d5b55105ab19ad773c3a795b2f1a3b052b1f6008.sig
 ```
 Get Public Key:
 ```
 cosign public-key --key awskms:///alias/cosign-ec384 > kms-ec384.pub
 ```
 Verify:
 ```
cosign verify --key ./kms-ec384.pub <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:ECDSA_SHA_384
 ```
 Verify Output:
 ```
error: no matching signatures:
failed to verify signature
 ```

AWS KMS 3 (alias/cosign-ec256):
-------------------------------
Key Type - Asymmetric
Origin: AWS_KMS
Key Spec: ECC_NIST_P256
Key Usage: Sign and Verify
Signing algorithms: ECDSA_SHA_256

Sign:
```
cosign sign --key awskms:///alias/cosign-ec256 <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:ECDSA_SHA_256
```
Sign Output:
 ```
 Pushing signature to: <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:sha256-445a2b673d8ec8937555ba159c721c4fe5d7f6fc0296ec97ce1f121826b58321.sig
 ```
 Get Public Key:
 ```
 cosign public-key --key awskms:///alias/cosign--ec256 > kms--ec256.pub
 ```
 Verify:
 ```
cosign verify --key ./kms--ec256.pub <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:ECDSA_SHA_256
 ```
 Verify Output:
 ```
Verification for <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:ECDSA_SHA_256 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - The signatures were verified against the specified public key
  - Any certificates were verified against the Fulcio roots.

[{"critical":{"identity":{"docker-reference":"<ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>"},"image":{"docker-manifest-digest":"sha256:445a2b673d8ec8937555ba159c721c4fe5d7f6fc0296ec97ce1f121826b58321"},"type":"cosign container image signature"},"optional":null}]
 ```

AWS KMS 4 (alias/cosign-rsa):
-------------------------------
Key Type - Asymmetric
Origin: AWS_KMS
Key Spec: RSA_4096
Key Usage: Sign and Verify
Signing algorithms: 
RSASSA_PKCS1_V1_5_SHA_256
RSASSA_PKCS1_V1_5_SHA_384
RSASSA_PKCS1_V1_5_SHA_512
RSASSA_PSS_SHA_256
RSASSA_PSS_SHA_384
RSASSA_PSS_SHA_512

Sign:
```
cosign sign --key awskms:///alias/cosign-rsa4096 <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:RSA_4096
```
Sign Output:
 ```
 Pushing signature to: <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:sha256-37edc48edc36ae72bc852081a0e55cd46e68d02e782a83569b0e67fb12a557d1.sig
 ```
 Get Public Key:
 ```
 cosign public-key --key awskms:///alias/cosign-rsa4096 > kms-rsa4096.pub
 ```
 Verify:
 ```
cosign verify --key ./kms-rsa.pub <ACCOUNT>.dkr.ecr.<REGION>.amazonaws.com/<REPO>/<IMAGE_NAME>:RSA_4096
 ```
 Verify Output:
 ```
error: loading public key: pem to ecdsa: invalid public key: was *rsa.PublicKey, require *ecdsa.PublicKey
 ```
 
- From the above experimentation, it seems only Elliptic curve based keys and algorithm SHA256 is only supported for validation. 

###  Expected Behavior:

If only Elliptic curve SHA 256 algorithm is supported. why allowing signing of images with all other format of keys and later fail during verify stage? Wouldn't it make sense to fail at Sign stage itself ?

### Improvements:

KMS.md can be improved with such key information as the supported type. 

I would appreciate if there is a confirmation on the above findings. Thanks in advance.

**Comment by developer-guy** [comment]
kindly ping @dlorenc @Dentrax @hectorj2f

**Comment by SgtCoDFish** [comment]
I can also confirm this kind of issue exists in GCP KMS for 4096-bit RSA keys with PKCS#1v1.5 padding and SHA512 digests.

Cosign will happily sign using such a key and **also verify using that key** if getting the key directly from GCP:

```console
$ cosign sign --key "gcpkms://..." ghcr.io/sgtcodfish/ace:0.5
$ cosign verify --key "gcpkms://..." ghcr.io/sgtcodfish/ace:0.5

Verification for ghcr.io/sgtcodfish/ace:0.5 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - The signatures were verified against the specified public key
  - Any certificates were verified against the Fulcio roots.

[{"critical":{"identity":{"docker-reference":"ghcr.io/sgtcodfish/ace"},"image":{"docker-manifest-digest":"sha256:292032005d1316d52f561487e20fccf0ccd366de1490ee4321f783a83e45e1d4"},"type":"cosign container image signature"},"optional":null}]
```

But with that key exported to a [local file](https://github.com/SgtCoDFish/ace/blob/f47ab3817e1ce8ee30a2221e225b273c19738916/PUBKEY), the verification fails:

```console
$ cosign verify --key ./PUBKEY ghcr.io/sgtcodfish/ace:0.5
Error: no matching signatures:
crypto/rsa: verification error
main.go:48: error during command execution: no matching signatures:
crypto/rsa: verification error
```

The issue here is the digest algorithm; when verifying using the `gcpkms://` reference directly, the verifier can deduce that it needs a SHA512 hash algorithm because that information is baked into to the GCP KMS key. When loading from a local PEM file the hash algorithm is [hardcoded and not configurable](https://github.com/sigstore/cosign/blob/5d866c373cfc983c2ec4455f36bc6388f64e9aaa/pkg/signature/keys.go#L56). I've confirmed locally that by changing that hardcoded hash algo to SHA512 it validates the above signature correctly.

One quick fix for this would be to allow the hash algorithm to be specified as an option, allowing me to provide the required context which cosign needs, e.g.

```console
$ cosign verify --key ./PUBKEY --hash-algorithm "SHA512" ghcr.io/sgtcodfish/ace:0.5
```

In addition there are heuristics which could be useful here; if the signing key is RSA4096 or P-512, it's common in my experience that SHA512 would be used as the hash algorithm (and likewise for P-384 it'd be common to use SHA384); in other words, I'd expect the default when faced with a 4096-bit RSA key or a P-521 ECDSA key should be to assume SHA512 unless someone passes a different `--hash-algorithm` explicitly.

I'd be interested in taking a swing at adding the `--hash-algorithm` flag to `cosign verify` (and `verify-blob`, I guess?), and / or trying to add heuristics for working out which hash algorithm to default to, if that would be appreciated; also happy to discuss further!

(Context for all this: in [cert-manager](https://github.com/jetstack/cert-manager) we _really_ want to have exactly one GCP KMS signing key for all artifacts, which to start would be our helm charts and our container images. Helm seems tightly tied to RSA + SHA512 in a very difficult-to-change way, so our hands are tied on SHA512 and that's how we've generated our signing key. Cosign is sort of the opposite where it assumes SHA256, which leaves us in a bit of a catch-22 situation currently.)
